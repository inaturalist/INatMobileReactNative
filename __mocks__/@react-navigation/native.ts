// 20240806 amanda - best practice for react navigation
// is actually not to mock navigation at all. I removed
// useFocusEffect so we can test how that actually works in
// components; it requires using the wrapInNavigationContainer
// helper around components with useFocusEffect
// https://reactnavigation.org/docs/testing/#best-practices

const actualNav = jest.requireActual( "@react-navigation/native" );

const useRoute = jest.fn( ( ) => ( { params: {} } ) );

const useNavigation = ( ) => ( {
  addListener: jest.fn( ),
  canGoBack: jest.fn( ( ) => true ),
  goBack: jest.fn( ),
  setOptions: jest.fn( )
} );

module.exports = {
  ...actualNav,
  useNavigation,
  useRoute
};
