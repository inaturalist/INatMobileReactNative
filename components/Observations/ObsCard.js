// @flow

import React from "react";
import { Pressable, Text, View, Image } from "react-native";
import type { Node } from "react";

import { viewStyles, textStyles } from "../../styles/observations/obsCard";

type Props = {
  item: {
    uuid: string,
    userPhoto: string,
    commonName: string,
    location: string,
    timeObservedAt: string,
    identifications: number,
    comments: number,
    qualityGrade: string
  },
  handlePress: Function
}

const ObsCard = ( { item, handlePress }: Props ): Node => (
  <Pressable
    onPress={handlePress}
    style={viewStyles.row}
    testID="ObsList.obsCard"
  >
    <Image source={{ uri: item.userPhoto }} style={viewStyles.imageBackground} />
    <View style={viewStyles.obsDetailsColumn}>
      <Text style={textStyles.text}>{item.commonName}</Text>
      <Text style={textStyles.text}>{item.location}</Text>
      <Text style={textStyles.text}>{item.timeObservedAt}</Text>
    </View>
    <View>
      <Text style={textStyles.text}>{item.identifications}</Text>
      <Text style={textStyles.text}>{item.comments}</Text>
      <Text style={textStyles.text}>{item.qualityGrade}</Text>
    </View>
  </Pressable>
);

export default ObsCard;
