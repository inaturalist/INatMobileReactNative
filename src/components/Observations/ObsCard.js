// @flow

import React from "react";
import { Pressable, Text, View, Image } from "react-native";
import type { Node } from "react";

import { viewStyles, textStyles } from "../../styles/observations/obsCard";

type Props = {
  item: Object,
  handlePress: Function
}

const ObsCard = ( { item, handlePress }: Props ): Node => (
  <Pressable
    onPress={( ) => handlePress( item )}
    style={viewStyles.row}
    testID={`ObsList.obsCard.${item.uuid}`}
    accessibilityRole="link"
    accessibilityLabel="Navigate to observation details screen"
  >
    <Image
      source={{ uri: item.observationPhotos[0].photo.url }}
      style={viewStyles.imageBackground}
      testID="ObsList.photo"
    />
    <View style={viewStyles.obsDetailsColumn}>
      <Text style={textStyles.text}>{item.taxon.preferredCommonName}</Text>
      <Text style={textStyles.text}>{item.placeGuess}</Text>
      <Text style={textStyles.text}>{item.timeObservedAt}</Text>
    </View>
    <View>
      <Text style={textStyles.text}>{item.identifications.length}</Text>
      <Text style={textStyles.text} testID="ObsList.obsCard.commentCount">{item.comments.length}</Text>
      <Text style={textStyles.text}>{item.qualityGrade}</Text>
    </View>
  </Pressable>
);

export default ObsCard;
