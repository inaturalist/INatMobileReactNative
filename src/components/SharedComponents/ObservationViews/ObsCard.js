// @flow

import { Image, Pressable, View } from "components/styledComponents";
import { RealmContext } from "providers/contexts";
import type { Node } from "react";
import React, { useEffect, useState } from "react";
import { Avatar } from "react-native-paper";

import Observation from "../../../models/Observation";
import Photo from "../../../models/Photo";
import ObsCardDetails from "./ObsCardDetails";
import ObsCardStats from "./ObsCardStats";

const { useRealm } = RealmContext;

type Props = {
  item: Object,
  handlePress: Function
}

const ObsCard = ( { item, handlePress }: Props ): Node => {
  const [needsUpload, setNeedsUpload] = useState( false );
  const onPress = ( ) => handlePress( item );
  const realm = useRealm( );

  const photo = item?.observationPhotos?.[0]?.photo;

  useEffect( ( ) => {
    const markAsNeedsUpload = async ( ) => {
      const isUnsyncedObs = Observation.isUnsyncedObservation( realm, item );
      setNeedsUpload( isUnsyncedObs );
    };
    markAsNeedsUpload( );
  }, [item, realm] );

  return (
    <Pressable
      onPress={onPress}
      className="flex-row my-2 mx-3 justify-between"
      testID={`ObsList.obsCard.${item.uuid}`}
      accessibilityRole="link"
      accessibilityLabel="Navigate to observation details screen"
    >
      <View className="flex-row shrink">
        <Image
          source={{ uri: Photo.displayLocalOrRemoteSquarePhoto( photo ) }}
          className="w-16 h-16 rounded-md mr-2"
          testID="ObsList.photo"
        />
        <View className="shrink">
          <ObsCardDetails item={item} />
        </View>
      </View>
      <View className="flex-row items-center justify-items-center ml-2">
        {needsUpload
          ? <Avatar.Icon size={40} icon="arrow-up-circle-outline" />
          : <ObsCardStats item={item} type="list" />}
      </View>
    </Pressable>
  );
};

export default ObsCard;
