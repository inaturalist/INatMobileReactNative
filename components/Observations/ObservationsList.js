// @flow strict-local

import React from "react";
import { FlatList, SafeAreaView } from "react-native";
import type { Node } from "react";

import ObsCard from "./ObsCard";
import viewStyles from "../../styles/observations/observationsList";
// import useFetchObservations from "./hooks/fetchObservations";
import EmptyList from "./EmptyList";
import useFetchLocalObservations from "./hooks/fetchLocalObservations";
import ViewWithFooter from "../SharedComponents/ViewWithFooter";

const ObservationsList = ( ): Node => {
  const localObservations = useFetchLocalObservations( );
  // this custom hook fetches on first component render
  // (and anytime you save while in debug - hot reloading mode )
  // useFetchObservations( );

  const handlePress = ( ) => console.log( "nav to obs details from obs list" );
  const extractKey = item => item.uuid;
  const renderItem = ( { item } ) => <ObsCard item={item} handlePress={handlePress} />;

  const renderEmptyState = ( ) => <EmptyList />;

  return (
    <ViewWithFooter>
      <SafeAreaView style={viewStyles.safeAreaContainer}>
        <FlatList
          data={localObservations}
          keyExtractor={extractKey}
          renderItem={renderItem}
          testID="ObservationsList.myObservations"
          ListEmptyComponent={renderEmptyState}
        />
      </SafeAreaView>
    </ViewWithFooter>
  );
};

export default ObservationsList;
