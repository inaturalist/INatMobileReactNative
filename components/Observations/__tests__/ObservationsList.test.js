import React from "react";
import { render } from "@testing-library/react-native";
import { NavigationContainer } from "@react-navigation/native";

import ObservationsList from "../ObservationsList";
import ObsCard from "../ObsCard";

test( "it renders all inputs as expected", ( ) => {
  const { toJSON } = render(
    // need nav container here so the useNavigation hook in Footer does not fail
    // not sure if there's a better way of doing this without needing to wrap every test
    // https://www.reactnativeschool.com/setup-jest-tests-with-react-navigation
    <NavigationContainer>
      <ObservationsList />
    </NavigationContainer>
  );
  expect( toJSON ).toMatchSnapshot( );
} );

test( "renders text passed into observation card", ( ) => {
  const { getByTestId, getByText } = render(
    <ObsCard
      item={{
        commonName: "Insects",
        location: "SF"
      }}
    />
  );

  expect( getByTestId( "ObservationsList.obsCard" ) ).toBeTruthy( );
  expect( getByText( "Insects" ) ).toBeTruthy( );
  expect( getByText( "SF" ) ).toBeTruthy( );
} );
