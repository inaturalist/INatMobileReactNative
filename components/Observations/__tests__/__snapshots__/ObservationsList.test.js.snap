// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders all inputs as expected 1`] = `[Function]`;
